### DISCUSSION TITLE
Code commenting

### DISCUSSION THREAD

**Discussion by PSTB** [Koenkk/zigbee2mqtt#16330] [General]
First of all, this is just out of curiosity, no criticism meant in any way. This software is absolutely great and helps me/us immensely! Many thanks to Koen et al. for it. 

I'm struggling throught the code to try to understand how Z2M works to help support some devices, but I feel there are awfully few comments in the code to document it. Is it just me or is it really the case ?

**Comment by Koenkk** [comment]
There are docs available on how to add support for new devices.
- generic: https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html
- tuya: https://www.zigbee2mqtt.io/advanced/support-new-devices/02_support_new_tuya_devices.html

**Reply by PSTB** [reply]
Yes Koen, and I definitely _have_ read those. 
But when reading through the code to get a better understanding I was really surprised by the sparse comments.
As a js/ts beginner I was wondering if this is usual...
